content/doc/book/pipeline/pipeline-best-practices.adoc

```diff
@@ -0,0 +1,95 @@
+Pipelines, as their complexity increases (the amount of Groovy code, number of steps used, etc.), require more resources (CPU, memory, storage) on the master.
+Think of Pipeline as a tool to accomplish a build rather than the core of a build.
+
+Example: Using a single Maven build step to drive the build through its build/test/deploy process.
+
+=== Avoiding complex Groovy code in Pipelines
+
+For a Pipeline, Groovy code *always* executes on master which means using master resources(memory and CPU). 
+Therefore, it is critically important to reduce the amount of Groovy code executed by Pipelines (this includes any methods called on classes imported in Pipelines).
+The following are the most common example Groovy methods to avoid using:
+. *JsonSlurper:* This function (and some other similar ones like XmlSlurper or readFile) can be used to read from a file on disk, parse the data from that file into a JSON object, and inject that object into a Pipeline using a command like JsonSlurper().parseText(readFile("$LOCAL_FILE")). This command loads the local file into memory on the master twice and, if the file is very large or the command is executed frequently, will require a lot of memory.
+.. Solution: Instead of using JsonSlurper, use a shell step and return the standard out. This shell would look something like this: `def JsonReturn = sh label: '', returnStdout: true, script: 'echo "$LOCAL_FILE"| jq "."'`. This will use agent resources to read the file and even parse the file with a good `jq` filter
+. *HttpRequest:* Frequently this command is used to grab data from an external source and store it in a variable. This practice is not ideal because not only is that request coming directly from the master (which could give incorrect results for things like HTTPS requests if the master does not have certificates loaded), but also the response to that request is stored twice.
+.. Solution: Use a shell step to perform the HTTP request from the agent, for example using a tool like `curl` or `wget`, as appropriate. If the result must be later in the Pipeline, try to filter the result on the agent side as much as possible so that only the minimum required information must be transmitted back to the Jenkins master.
+
+=== Reducing repetition of similar Pipeline steps
+
+Combine Pipeline steps into single steps as often as possible to reduce the size of the flow-node graph(Pipeline’s way of tracking what step you are currently on). 
+Combine Pipeline steps into single steps as much as possible to reduce the amount of overhead caused by the Pipeline execution engine itself. For example, if you run three shell steps back-to-back, each of those steps has to be started and stopped, requiring connections and resources on the agent and master to be created and cleaned up. However, if you put all of the commands into a single shell step, then only a single step needs to be started and stopped.
```

Comment:
```suggestion
This also reduces the amount of overhead caused by the Pipeline execution engine itself. For example, if you run three shell steps back-to-back, each of those steps has to be started and stopped, requiring connections and resources on the agent and master to be created and cleaned up. However, if you put all of the commands into a single shell step, then only a single step needs to be started and stopped.
```

Reply:
Good catch. FWIW, I would get rid of the dupe by deleting the first sentence, since the size of the flow graph is not something users should care about except as a proxy for other things:

> Combine Pipeline steps into single steps as often as possible to reduce the size of the flow-node graph(Pipeline’s way of tracking what step you are currently on).